### PR TITLE
feat(payment): INT-4222 Add vaulting compatibility to Moneris

### PR DESCRIPTION
**Pull request by FranciscoOnt**
## What? [(INT-4222)](https://jira.bigcommerce.com/browse/INT-4222)
Update strategy to support vaulted instruments and add Moneris to supported instruments.

## Why?
So vaulted instruments can be used with Moneris.

## Testing / Proof
UInit Testing

@bigcommerce/checkout @bigcommerce/payments @bigcommerce/apex-integrations

### CODE REVIEW

**Comment by JoseBobadilla20** [comment]
💚 Tested 
QA Details: I check that the Moneris vaulting flow works correctly.

File: src/payment/strategies/moneris/moneris-payment-strategy.ts
Comment (vmparra) [inline src/payment/strategies/moneris/moneris-payment-strategy.ts:60]: You should first validate that instrument is in fact a nonce. See https://github.com/bigcommerce/checkout-sdk-js/blob/master/src/payment/is-nonce-like.ts

**Review by vmparra — CHANGES_REQUESTED** [review]
Hi @FranciscoOnt So far looks real good, just left one comment. Please let me know if you have questions.

File: src/payment/strategies/moneris/moneris-payment-strategy.ts
Comment (JoseAlvarezHdz) [inline src/payment/strategies/moneris/moneris-payment-strategy.ts:97]: i think you are missing the `shouldSetAsDefaultInstrument` param 🤔

File: src/payment/strategies/moneris/moneris-payment-strategy.ts
Comment (FranciscoOnt) [reply src/payment/strategies/moneris/moneris-payment-strategy.ts:60]: I don't think that will work, the `shouldSaveIntrument` is received when the payment is not yet a nonce instrument and since the `isNonceLike` validates the `nonce` field which is missing from this initial payload it will alway return false in the Moneris scenario then making the `shouldSaveIntrument`always undefined.

Comment (FranciscoOnt) [reply src/payment/strategies/moneris/moneris-payment-strategy.ts]: Just to clarify that, I changed to the `HostedInstrument` Type

**Review by JoseAlvarezHdz — APPROVED** [review]
LGTM 👍🏾

File: src/payment/strategies/moneris/moneris-payment-strategy.ts
Comment (vmparra) [inline src/payment/strategies/moneris/moneris-payment-strategy.ts:116]: You're still missing the type guard. You should also cast `paymentData` to HostedInstrument only once.

**Review by vmparra — CHANGES_REQUESTED** [review]
It looks real good however is would be wise to have a type guard for HostedInstrument and then cast the object.

**Review by vmparra — APPROVED** [review]
Great work @FranciscoOnt 

image

**Review by JoseAlvarezHdz — APPROVED** [review]
LGTM! 👍🏾  nice!

**Comment by JoseBobadilla20** [comment]
💚 Tested
QA Details: I check that the Moneris vaulting flow and TVS flow works correctly.

**Review by Tharaae — COMMENTED** [review]
Generally LGTM, Good job 👍 
But please add the testing proof to the PR description while I have another more detailed look.

**Review by Tharaae — APPROVED** [review]
Checked the testing proof on the sibling checkout-js PR. LGTM 👍

**Review by Tharaae — APPROVED** [review]

File: src/payment/strategies/moneris/moneris-payment-strategy.ts
Comment (animesh1987) [inline src/payment/strategies/moneris/moneris-payment-strategy.ts:168]: This method should be prefixed with a `_`.

File: src/payment/strategies/moneris/moneris-payment-strategy.ts
Comment (animesh1987) [inline src/payment/strategies/moneris/moneris-payment-strategy.ts:113]: Why is this a protected method and not a private? and can we please add a `_` as a prefix.

File: src/payment/strategies/moneris/moneris-payment-strategy.ts
Comment (animesh1987) [inline src/payment/strategies/moneris/moneris-payment-strategy.ts:151]: Why is this a protected method and not a private? and can we please add a _ as a prefix.

File: src/payment/strategies/moneris/moneris-payment-strategy.ts
Comment (animesh1987) [inline src/payment/strategies/moneris/moneris-payment-strategy.ts:162]: Why do we need to load the current order here?

File: src/payment/strategies/moneris/moneris-payment-strategy.ts
Comment (animesh1987) [inline src/payment/strategies/moneris/moneris-payment-strategy.ts:193]: The new promise construction is redundant, you can move `async` keyword on `_mountCardVerificationfields ` function and it should be good.

File: src/payment/strategies/moneris/moneris-payment-initialize-options.ts
Comment (animesh1987) [inline src/payment/strategies/moneris/moneris-payment-initialize-options.ts:16]: Seems like the change you are doing here is to use the hosted form fields for vaulted instrument verification and the `HostedFormOptions` accepts parameters more than just instrument verification. For reference you can see this https://github.com/bigcommerce/checkout-sdk-js/blob/master/src/hosted-form/hosted-form-options.ts#L20. As far as I can tell if the client passes those options it will not work. But here you are looking for https://github.com/bigcommerce/checkout-sdk-js/blob/master/src/hosted-form/hosted-form-options.ts#L29 this particular interface to be accepted.

So maybe you need to describe a new interface to check for the supplied parameters?

File: src/payment/strategies/moneris/moneris-payment-strategy.ts
Comment (FranciscoOnt) [reply src/payment/strategies/moneris/moneris-payment-strategy.ts:162]: It's used to update the state of the checkout and check if the payment for the order has been submitted, neither `form.validate` and `form.submit` update the state.
This is what the base credit-card strategy does.

File: src/payment/strategies/moneris/moneris-payment-strategy.ts
Comment (FranciscoOnt) [reply src/payment/strategies/moneris/moneris-payment-strategy.ts:193]: if the promise isn't marked as async, the line with `await form.attach();`  will be marked as an error, since the arrow function make it be in a different scope that would no longer be async.

**Comment by FranciscoOnt** [comment]
@animesh1987 Could you please re-review ♻️

File: src/payment/strategies/moneris/moneris-payment-strategy.ts
Comment (animesh1987) [reply src/payment/strategies/moneris/moneris-payment-strategy.ts:193]: Have you tried it removing the promise and putting `async` next to `_mountCardVerificationfields `?

Comment (FranciscoOnt) [reply src/payment/strategies/moneris/moneris-payment-strategy.ts:193]: yes, that's what I tried, but it didn't work

**Comment by PascalZajac** [comment]
Looks like this still needs final review from @bigcommerce/checkout before it's actually ready to merge.

File: src/payment/strategies/moneris/moneris-payment-strategy.ts
Comment (animesh1987) [reply src/payment/strategies/moneris/moneris-payment-strategy.ts:193]: Have you tried this 
```
private async _mountCardVerificationfields(formOptions: HostedFormOptions): Promise<HostedForm> {
        try {
            const { config } = this._store.getState();
            const bigpayBaseUrl = config.getStoreConfig()?.paymentSettings.bigpayBaseUrl;

            if (!bigpayBaseUrl) {
                throw new MissingDataError(MissingDataErrorType.MissingCheckoutConfig);
            }

            const form = this._hostedFormFactory.create(bigpayBaseUrl, formOptions);

            await form.attach();

            return form;
        } catch (error) {
            return error;
        }
    }
```

Comment (FranciscoOnt) [reply src/payment/strategies/moneris/moneris-payment-strategy.ts]: that does, work, I've added it to the PR

File: src/payment/strategies/moneris/moneris-payment-strategy.ts
Comment (animesh1987) [inline src/payment/strategies/moneris/moneris-payment-strategy.ts:193]: Here we accept `HostedFormOptions` but `monerisOptions.form` is a different interface `HostedFormValidationOptions`. So maybe we need to adjust the types in all relevant places.

File: src/payment/strategies/moneris/moneris-payment-strategy.ts
Comment (animesh1987) [reply src/payment/strategies/moneris/moneris-payment-strategy.ts]: Thanks for updating that.

**Review by animesh1987 — APPROVED** [review]

**Pull request merged**
